bus-mapping/src/circuit_input_builder/execution.rs

```diff
@@ -365,47 +397,79 @@ pub struct CopyEvent {
     pub log_id: Option<u64>,
     /// Value of rw counter at start of this copy event
     pub rw_counter_start: RWCounter,
-    /// Represents the list of (bytes, is_code) copied during this copy event
-    pub bytes: Vec<(u8, bool)>,
+    /// Represents the list of bytes related during this copy event
+    pub copy_bytes: CopyBytes,
 }
 
+pub type CopyEventSteps = Vec<(u8, bool, bool)>;
+pub type CopyEventPrevBytes = Vec<u8>;
+
 impl CopyEvent {
     /// rw counter at step index
     pub fn rw_counter(&self, step_index: usize) -> u64 {
         u64::try_from(self.rw_counter_start.0).unwrap() + self.rw_counter_increase(step_index)
     }
 
+    /// rw counter at step index
+    pub fn rw_counter_step(&self, step_index: usize) -> u64 {
+        let mut rw_counter = u64::try_from(self.rw_counter_start.0).unwrap();
+        let rw_counter_increase = self.rw_counter_increase(step_index);
+        rw_counter += rw_counter_increase;
+
+        // step_index == self.bytes.len() when caculate total rw increasing.
+        if self.dst_type == CopyDataType::TxLog
+            && step_index != self.copy_bytes.bytes.len() * 2
+            && step_index % 64 == 63
+        {
+            // log writing
+            rw_counter += 1;
+        }
+
+        rw_counter
+    }
+
     /// rw counter increase left at step index
     pub fn rw_counter_increase_left(&self, step_index: usize) -> u64 {
-        self.rw_counter(self.bytes.len() * 2) - self.rw_counter(step_index)
+        if self.rw_counter_step(self.copy_bytes.bytes.len() * 2) < self.rw_counter_step(step_index)
+        {
+            panic!("prev rw_counter_step > total tw_counter");
+        }
+        // self.rw_counter_step(self.bytes.len() * 2) - self.rw_counter_step(step_index)
+        self.rw_counter_step(self.copy_bytes.bytes.len() * 2) - self.rw_counter_step(step_index)
     }
 
     /// Number of rw operations performed by this copy event
     pub fn rw_counter_delta(&self) -> u64 {
-        self.rw_counter_increase(self.bytes.len() * 2)
+        self.rw_counter_increase(self.copy_bytes.bytes.len() * 2)
     }
 
     // increase in rw counter from the start of the copy event to step index
     fn rw_counter_increase(&self, step_index: usize) -> u64 {
+        if let (CopyDataType::Memory, CopyDataType::Memory) = (self.src_type, self.dst_type) {
+            return step_index as u64 % 2 + 2 * (step_index as f32 / 64.0).floor() as u64;
```

Comment:
```
return step_index as u64 % 2 + 2 * (step_index as u64 / 64);
```
?

Reply:
refactored new common helper [refactor write_chunks&write_chunk_for_copy_step](https://github.com/scroll-tech/zkevm-circuits/pull/466/commits/ba6a42c0252212b87d823e3e4c9c67f4cc47fab6)

Reply:
@lightsing  this is added by handling memory-->memory case, can you explain it ?